### PR TITLE
refactor(api): add custom labware to legacy protocol runs

### LINKED ISSUE #8525: HTTP API: allow custom labware files to be uploaded with protocol resource
## Overview

As a user of the HTTP API, I would like any custom labware files I upload to `/protocols` to be available in my Python protocol.

For the engine-based `POST /protocols` endpoint, we should analyze any JSON files in the request, and identify if they are labware definitions. Then, if that protocol resource is used to create a run, those definitions should be added to the protocol engine

## Acceptance criteria

- [ ] Can `POST /protocols` with a Python file and multiple labware definition JSON files
- [ ] Pre-analyzer can still identify the Python file as the "main" file
- [ ] Pre-analyzer identifies definition files as `LabwareDefinition`'s
- [ ] `ProtocolRunner` can read those definition files at protocol load and put them into engine state

### PR DESCRIPTION
**Pull request by mcous**
## Overview

Closes #8525.

This PR adds custom labware support to the `ProtocolReader` and `ProtocolRunner`.

## Changelog

- `ProtocolReader` can now read/parse custom labware files
- `ProtocolRunner` will feed labware definitions from ProtocolReader into legacy protocol context implementation
    - This allows protocols to run with custom labware
- `ProtocolRunner` will add labware definitions to `ProtocolEngine` state
    - This allows HTTP protocol commands to load custom labware during LPC

The custom labware support in PE added by this PR also made the `addLabwareDefinition` PE command obsolete, so I removed it given nothing would be using it. I think as a command, `addLabwareDefinition` didn't quite fit in; it loaded no data, made no hardware calls, but simply added state to PE.

## Review requests

Besides typical code review, here are the scenarios I think we need to smoke test:

- LPC + run works with Python protocol that includes custom labware files
    - Note: _all_ custom labware files uploaded automatically by the app will appear in the file list of the `/protocols` response, regardless of whether the Python protocol actually uses them
- LPC + run works with JSON protocol that includes integrated custom labware
    - Note: none of custom labware files uploaded automatically by the app will appear in the files list of the `/protocols` response, because they are not used with JSON protocols

## Risk assessment

Low.

### CODE REVIEW

File: api/src/opentrons/protocol_reader/protocol_source.py [hunk 1]
```diff
@@ -15,19 +16,20 @@ class ProtocolType(str, Enum):
     PYTHON = "python"
 
 
+# TODO(mc, 2021-12-07): add custom labware, data, and python support roles
```
Comment (mcous) [inline api/src/opentrons/protocol_reader/protocol_source.py]:
```suggestion
# TODO(mc, 2021-12-07): add data and python support roles
```

File: api/src/opentrons/protocol_reader/protocol_source.py [hunk 2]
```diff
@@ -74,21 +76,24 @@ class PythonProtocolConfig:
...
 
     Attributes:
-        path: The directory location of the protocol on disk.
-        files: The protocol's files.
-
+        directory: The directory location of the protocol on disk.
+        main_file: The location of the protocol's main file on disk.
+        files: Descriptions of all files that make up the protocol.
+        metadata: Arbitrary metadata specified by the protocols.
+        config: Protocol execution configuration.
```
Comment (mcous) [inline api/src/opentrons/protocol_reader/protocol_source.py:100]:
```suggestion
        config: Protocol execution configuration.
        labware: Custom labware definitions included with the protocol.
```

File: api/src/opentrons/protocol_runner/legacy_wrappers.py [hunk 3]
```diff
@@ -86,38 +95,61 @@ def __init__(
...
-            return LegacyProtocolContext(
+                hardware=self._hardware_api,
+                extra_labware={
+                    uri_from_details(
+                        namespace=lw.namespace,
+                        load_name=lw.parameters.loadName,
+                        version=lw.version,
+                    ): cast(LegacyLabwareDefinition, lw.dict(exclude_none=True))
+                    for lw in labware
+                },
```
Comment (mcous) [inline api/src/opentrons/protocol_runner/legacy_wrappers.py]: Move to helper function

File: api/src/opentrons/protocol_engine/state/labware.py [hunk 4]
```diff
@@ -116,6 +137,12 @@ def _handle_command(self, command: Command) -> None:
             )
 
             new_labware_by_id = self._state.labware_by_id.copy()
+
+            if command.params.location == _TRASH_LOCATION:
+                for target_id, target in self._state.labware_by_id.items():
+                    if target.location == _TRASH_LOCATION:
+                        del new_labware_by_id[target_id]
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/state/labware.py]: This logic deserves its own PRs, especially in light of #9062

**Comment by SyntaxColoring** [comment]
Tested on a robot. This appears to be working except for the client giving the server the full, absolute file path of labware definitions instead of just the filename, which causes a 500 error.

It's probably a client bug that that it's giving the server absolute paths, but it's also probably a server bug that we're not rejecting those absolute paths with a 422.

File: app/src/redux/custom-labware/__tests__/selectors.test.ts [hunk 5]
```diff
@@ -178,6 +151,38 @@ describe('custom labware selectors', () => {
 
   SPECS.forEach(spec => {
     const { name, selector, state, expected } = spec
-    it(name, () => expect(selector(state)).toEqual(expected))
+    it(`should handle ${name}`, () => expect(selector(state)).toEqual(expected))
+  })
+
+  it('should map custom labware to File blobs for upload', () => {
```
Comment (mcous) [inline app/src/redux/custom-labware/__tests__/selectors.test.ts:157]: Note: jest's `File` polyfill is quite bad. From what I can tell, it appears as simply `{}` during comparisons, so _any_ `File` will compare equal to any other `File`. Accordingly, I split this out into an explicit test because the specs-from-array wasn't going to be able to test this logic

**Review by smb2268 — APPROVED** [review]
Looks good - thanks for fixing!

File: api/src/opentrons/protocol_reader/file_reader_writer.py [hunk 6]
```diff
@@ -17,20 +36,31 @@ class FileReaderWriter:
...
+
             async with wrap_file(input_file.file) as f:
                 contents = await f.read()
                 data: Optional[BufferedJsonFileData] = None
 
                 if input_file.filename.endswith(".json"):
                     try:
-                        data = parse_raw_as(BufferedJsonFileData, contents)
-                    except (JSONDecodeError, ValidationError) as e:
+                        data = parse_raw_as(BufferedJsonFileData, contents)  # type: ignore[arg-type]  # noqa: E501
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_reader/file_reader_writer.py:52]: This doesn't necessarily have to happen in this PR, but we should treat parsing as long-running blocking work. We don't know how big and complicated the JSON file is.

So, either:

* This `read()` method should be `def`, and calling code should be expected to run it in a worker of some kind.
* Or it can remain `async def` and we can put it in a worker ourselves, here.

File: api/src/opentrons/protocol_reader/config_analyzer.py [hunk 7]
```diff
@@ -49,7 +49,7 @@ def _analyze_python(main_file: RoleAnalyzedFile) -> ConfigAnalysis:  # noqa: C90
         # don't truly want the protocol to inherit our own __future__ features.
         module_ast = ast.parse(source=main_file.contents, filename=main_file.name)
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_reader/config_analyzer.py:50]: Ditto the other comment on parsing being blocking work.

**Review by SyntaxColoring — APPROVED** [review]
Python changes mostly reviewed in a video call. 👍

**Review by sanni-t — APPROVED** [review]
Reviewed in call
